src/meltano/core/manifest/jsonschema.py

```diff
@@ -0,0 +1,129 @@
+"""Utilities to extract information from jsonschema files."""
+
+from __future__ import annotations
+
+from collections.abc import Iterable
+from contextlib import suppress
+from typing import Any
+
+
+def meltano_config_env_locations(
+    manifest_schema: dict[str, Any],
+    env_definition_ref: str = "#/$defs/env",
+) -> set[str]:
+    """Find all locations within a Meltano manifest file that can have an env field.
+
+    Args:
+        manifest_schema: The jsonschema for Meltano manifests.
+        env_definition_ref: The jsonschema ref that defines what is being searched for.
+
+    Returns:
+        The locations within a Meltano manifest file that can have an env
+        field, as jq filter strings.
+    """
+    return set(JsonschemaRefLocationParser(manifest_schema, env_definition_ref).parse())
+
+
+class JsonschemaRefLocationParser:
+    """Minimal jsonschema parser that finds the locations of a specified ref."""
+
+    def __init__(self, schema: dict[str, Any], target_ref: str) -> None:
+        """Initialize the `JsonschemaRefLocationParser`.
+
+        Args:
+            schema: The schema being parsed.
+            target_ref: The ref being searched for within the given schema.
+        """
+        self.root_schema = schema
+        self.target_ref = target_ref
+        self._resolved_refs: dict[str, Any] = {}
+
+    def parse(  # noqa: C901
+        self,
+        schema: dict[str, Any] | None = None,
+        path: tuple[str, ...] = (),
+    ) -> Iterable[str]:
```

Comment:
Might be worth being more specific on the return type

```suggestion
    ) -> Generator[str, None, None]:
```